source/2_platform/rex_windows/src/console_application.cpp

```diff
@@ -0,0 +1,114 @@
+          // see https://learn.microsoft.com/en-us/windows/console/handlerroutine#timeouts
+        case CTRL_CLOSE_EVENT:
+          g_this_app->quit();
+          return false; // NOLINT(readability-implicit-bool-conversion)
+          break;
+      }
+
+      return true; // NOLINT(readability-implicit-bool-conversion)
+    }
+
+    class ConsoleApplication::Internal
+    {
+    public:
+      Internal(CoreApplication* appInstance, ApplicationCreationParams&& appCreationParams)
+          : m_platform_creation_params(rsl::move(appCreationParams.platform_params))
+          , m_cmd_line_args(rsl::move(appCreationParams.cmd_args))
+          , m_engine_params(rsl::move(appCreationParams.engine_params))
+          , m_app_instance(appInstance)
+      {
+        g_this_app = m_app_instance;
+
+        // we're always assigning something to the pointers here to avoid branch checking every update
+        // I've profiled this and always having a function wins here.
+        m_on_initialize = m_engine_params.app_init_func ? m_engine_params.app_init_func : [&]() { return true; };
+
+        m_on_update = m_engine_params.app_update_func ? m_engine_params.app_update_func : [&]() {};
+
+        m_on_shutdown = m_engine_params.app_shutdown_func ? m_engine_params.app_shutdown_func : [&]() {};
+      }
+
+      bool initialize()
+      {
+        SetConsoleCtrlHandler(handler_routine, true); // NOLINT(readability-implicit-bool-conversion)
+
+        event_system::subscribe(event_system::EventType::QuitApp, [this](const event_system::Event& /*event*/) { m_app_instance->quit(); });
+
+        return m_on_initialize();
+      }
+
+      void update()
+      {
+        m_on_update();
+      }
+
+      void shutdown()
+      {
+        m_on_shutdown();
+      }
+
+      bool is_focussed() const // NOLINT(readability-convert-member-functions-to-static)
+      {
+        return GetConsoleWindow() == GetForegroundWindow();
+      }
+
+    private:
+      PlatformCreationParams m_platform_creation_params;
+      CommandLineArguments m_cmd_line_args;
+      EngineParams m_engine_params;
+      CoreApplication* m_app_instance;
+
+      rsl::function<bool()> m_on_initialize;
+      rsl::function<void()> m_on_update;
+      rsl::function<void()> m_on_shutdown;
+    };
+
+    ConsoleApplication::ConsoleApplication(ApplicationCreationParams&& appParams)
+        : CoreApplication(appParams.engine_params, appParams.cmd_args)
+        , m_internal_obj()
+        , m_internal(m_internal_obj.get<ConsoleApplication::Internal>())
+    {
+      static_assert(s_internal_size == sizeof(ConsoleApplication::Internal), "size of internal structure should match size of internal implementation");
+      static_assert(s_internal_alignment == alignof(ConsoleApplication::Internal), "alignment of internal structure should match alignment of internal implementation");
+
+      new(m_internal)(ConsoleApplication::Internal)(this, rsl::move(appParams));
```

Comment:
This really seems like overkill to me ... 
Why does this need to be this complex?
We only have 1 console app so the one allocation is not an issue here.
Readability should always triumph over verbosity only when this poses a real issue we should consider optimizing this.

Reply:
The reason why I made it stack and not heap memory is to avoid cluttering the memory tracking these kind of allocations.
Another way would be to use the untracked allocator

Reply:
Can't we add a tag for pimpl or internal? We have a memory tagging system why not use it?

Reply:
Yes we can use a tag for that